[WiFi] Improve WiFi reconnect stability and fix initial setup

TD-er opened the pull request:
Simplified the WiFi connect state machine.
Fixed setup of initial WiFi configuration.

AP mode is now kept as long as client is connected.
AP mode will be started immediately if there are no valid WiFi credentials, or after 20 seconds of no wifi connection.
AP mode will be disabled 60 seconds after starting it, or after the last client disconnected from the AP.

Fixes #2561 
Fixes #2338 
Fixes #2203 
Fixes #1987 
Fixes #1737
Fixes #1640 
Fixes #1565
Fixes #1479 
Fixes #1098
Fixes #558

TD-er committed:
[WiFi] Improve WiFi reconnect stability and fix initial setup

Simplified the WiFi connect state machine.
Fixed setup of initial WiFi configuration.

AP mode is now kept as long as client is connected.
AP mode will be started immediately if there are no valid WiFi credentials, or after 20 seconds of no wifi connection.
AP mode will be disabled 60 seconds after starting it, or after the last client disconnected from the AP.

TD-er commented:
Still to do:
- Captive Portal page (initial setup) is not working well
- Still some unneeded complexity left in the code.
- Setup page should be able to refresh the WiFi scan.
- Setup page should have an option to continue setup via AP (thus access / while connected to 192.168.4.1)
- Sometimes WiFi AP cannot be found for a longer time, forcing WIFI_OFF and restart WiFi may help in such situations.
- WiFi configuration should allow to use more APs (load from file?)
- Sort result of WiFi scan on signal strength.
- Show WiFi RSSI as icon or color.
- When connecting, perform a scan and then pick the strongest from the known SSIDs.
- Try to always use BSSID/channel  when connecting to AP, it is a lot faster.
- Cache last used BSSID/channel in the RTC memory to allow fast initial reconnect after (re)boot.
- Add setting to only start AP in some conditions (e.g. only in first N minutes after boot)
- Add setting to only start WiFi when some conditions are met (e.g. check Vcc first)
- Fix crash when continuously reading Vcc while connecting to WiFi. (oversampling of ADC plugin)

uzi18 commented on src/ESPEasy.ino:
```diff
@@ -1015,14 +1008,16 @@ void backgroundtasks()
 
   // process DNS, only used if the ESP has no valid WiFi config
-  if (dnsServerActive && wifiConnected)
+  if (dnsServerActive)
```
@TD-er all sort of wifi stuff should go to ESPEasyWifi.ino, what do you think?

TD-er replied:
Moving code around to the more logical locations was my next step indeed, but I wanted to do it in steps/commits, so it is still traceable.
But the one line you marked is just one that is not WiFi related and was for sure a bug that took me a while to pinpoint.
It was the source of reboots at initial WiFi setup and probably also causing crashes when there was no WiFi connection, but only the AP active.
What was happening here, is that the dns server (and web server) kept on adding stuff to their buffers unless they were being attended.
So you would run out of memory in minutes (or seconds if somehow requests came in)

I wouldn't be surprised if this also lead to crashes of nodes when they were trying to reconnect to WiFi.

uzi18 replied:
nice catch :)

TD-er committed:
[Docs] Document Timing Stats page

TD-er committed:
[WiFi] Move functions to separate files and document

Move WiFi/Network related functions into separate files to keep them organized and easier to locate.
Also added lots of comments and some documentation on the state machine handling WiFi connect and reconnects.

TD-er committed:
[WiFi] Clean up WiFi code and modernize WiFi setup page

The last page of the wizard now shows the network info, just like in the sysinfo page and has a button at the bottom which links to the IP address of the node on the normal network
This last page of the setup also has the menu bar at the top and also the default template now shows a warning when you're connected via AP mode.
This allows for easier setup, since you don't have to switch networks anymore

TD-er commented:
![image](https://user-images.githubusercontent.com/3751318/63627795-8b903580-c609-11e9-8a1e-38627e5f4866.png)

This is the last page of the (new) setup wizard.
It shows the network info, just like in the sysinfo page and has a button at the bottom which links to the IP address of the node on the normal network
This last page of the setup also has the menu bar at the top and also the default template now shows a warning when you're connected via AP mode.
This allows for easier setup, since you don't have to switch networks anymore

TD-er committed:
[LoRa] Keep out LoRa code if no supported controller is used

Building for core 2.6.0 was next to impossible with the LoRa related code to create packed raw data streams was included.

TD-er committed:
[WiFi] Add RSSI symbol to WiFi setup page

ghtester commented:
For me it would be great to have an option to disable the AP mode completely (from /config page).  Also for some devices that don't need to be permanently connected but are expected to work "more realtime" (I mean for instance  IRRX/ IRTX related plugins or some other where the timing is a bit critical) I would appreciate the possibility to configure node to try access the configured AP(s) only several times at boot.

TD-er commented:
@ghtester Something like this issue I created yesterday: https://github.com/letscontrolit/ESPEasy/issues/2565
But then for rules to start AP mode?

ghtester commented:
For me it's not important but what about modifying the Reset Pin to select "AP mode start" or "Factory defaults"?

TD-er commented:
> For me it's not important but what about modifying the Reset Pin to select "AP mode start" or "Factory defaults"?

Not sure how you plan to modify the reset pin to trigger different behavior in software.

ghtester commented:
Well, I don't know how the Reset Pin feature (on /hardware web page) exactly works but I thought it could be modified (listbox selection ?) to do factory reset or AP mode start instead.

TD-er commented:
That's an extra setting to allow to use another GPIO to also trigger a factory reset.

ghtester commented:
Or just one GPIO with selection what it should do. Just a thought... ;-)
